src/core/public/application/application_service.tsx

```diff
@@ -17,159 +17,157 @@
   private readonly capabilities = new CapabilitiesService();
+  private currentAppId$ = new BehaviorSubject<string | undefined>(undefined);
+  private registrationClosed = false;
+  private history: History<any> | null = null;
   private mountContext?: IContextContainer<App['mount']>;
+  private navigate?: (url: string, state: any) => void;
+
+  public setup({
+    context,
+    http: { basePath },
+    injectedMetadata,
+    redirectTo = (path: string) => (window.location.href = path),
+  }: SetupDeps): InternalApplicationSetup {
+    // Only setup history if we're not in legacy mode
+    if (!injectedMetadata.getLegacyMode()) {
+      this.history = createBrowserHistory({ basename: basePath.get() });
+    }
 
-  public setup({ context }: SetupDeps): InternalApplicationSetup {
+    // If we do not have history available, use redirectTo to do a full page refresh.
+    this.navigate = (url, state) =>
+      this.history ? this.history.push(url, state) : redirectTo(basePath.prepend(url));
     this.mountContext = context.createContextContainer();
 
     return {
-      register: (plugin: symbol, app: App) => {
-        if (this.apps$.value.has(app.id)) {
-          throw new Error(`An application is already registered with the id "${app.id}"`);
-        }
-        if (this.apps$.isStopped) {
+      registerMountContext: this.mountContext!.registerContext,
+      register: (plugin, app) => {
+        app = { appBasePath: `/app/${app.id}`, ...app };
+
+        if (this.registrationClosed) {
           throw new Error(`Applications cannot be registered after "setup"`);
+        } else if (this.apps.has(app.id)) {
+          throw new Error(`An application is already registered with the id "${app.id}"`);
+        } else if (hasAppBasePath(this.mounters, app.appBasePath)) {
+          throw new Error(
+            `An application is already registered with the appBasePath "${app.appBasePath}"`
+          );
         }
 
-        const appBox: AppBox = {
-          app,
-          mount: this.mountContext!.createHandler(plugin, app.mount),
+        const handler = this.mountContext!.createHandler(plugin, app.mount);
+        const mount: AppMounter = async params => {
+          const unmount = await handler(params);
+          this.currentAppId$.next(app.id);
+          return unmount;
         };
-        this.apps$.next(new Map([...this.apps$.value.entries(), [app.id, appBox]]));
+        this.apps.set(app.id, app);
+        this.mounters.set(app.id, {
+          appBasePath: app.appBasePath!,
+          mount,
+          unmountBeforeMounting: false,
+        });
       },
-      registerLegacyApp: (app: LegacyApp) => {
-        if (this.legacyApps$.value.has(app.id)) {
-          throw new Error(`A legacy application is already registered with the id "${app.id}"`);
-        }
-        if (this.legacyApps$.isStopped) {
+      registerLegacyApp: app => {
+        const appBasePath = `/app/${app.id.split(':')[0]}`;
```

Comment:
seems that all the functional tests are failing due to this. id format in LP is `kibana:{appId}`

Reply:
I think I already have this fixed locally, but I'll try it out on my next push.